### PR TITLE
Feature/2016 12 11 reactive

### PR DESCRIPTION
**Pull request by tomvdbulck**
blogpost about Spring Reactor

### CODE REVIEW

File: _posts/2016-12-12-Reactive-Programming-Spring-Reactor.md [hunk 1]
```diff
@@ -0,0 +1,525 @@
...
+* [Other Changes](#changes)
+* [RxJava](#rxjava)
+* [Spring Framework 5](#spring-5)
+* [Conclusion & Do It Yourself](#conclusion)
+
+<a name="join" />
+
+# Stephane Maldini @ JOIN 2016
+
+On 5 October 2016 we had the pleasure to welcome Stephane Maldini at our [JOIN event](https://ordina-jworks.github.io/conferences/2016/09/27/JOIN-2016.html)
```
Comment (bart-blommaerts) [inline _posts/2016-12-12-Reactive-Programming-Spring-Reactor.md]: +On 5 October 2016, we

File: _posts/2016-12-12-Reactive-Programming-Spring-Reactor.md [hunk 2]
```diff
@@ -0,0 +1,525 @@
...
+
+<img class="float-image p-image" src="{{ '/img/stephane-maldini.jpeg' | prepend: site.baseurl }}" alt="Stephane Maldini" style="width: 100px"/>
+
+A multi-tasker eating tech 24/7, Stephane is interested in cloud computing, data science and messaging.
+Leading the Reactor Project, Stephane Maldini is on a mission to help developers create reactive and efficient architectures on the JVM and beyond.
+He is also one of the main contributors for Reactive support in the upcoming Spring 5 framework.
+
+Which can be seen as the new standard for reactive applications in the Java world
```
Comment (bart-blommaerts) [inline _posts/2016-12-12-Reactive-Programming-Spring-Reactor.md]: This looks odd ...
Maybe ".. framework, which ..." and a "." at the end

File: _posts/2016-12-12-Reactive-Programming-Spring-Reactor.md [hunk 3]
```diff
@@ -0,0 +1,525 @@
...
+
+<a name="the-new-normal" />
+
+# The new normal that is not new
+
+It has been around for 30-40 years and boils down to [Event-Driven Programming](https://en.wikipedia.org/wiki/Event-driven_programming)
+What is new is "reactive motion bound to specification", this means that reactive programming is based on something solid, a specification and no longer some functional concepts.
+Namely the [Reactive Manifesto](http://www.reactivemanifesto.org/)
+
+Because of this specification, Spring found it the right time to start with [Reactor](https://spring.io/blog/2013/05/13/reactor-a-foundation-for-asynchronous-applications-on-the-jvm) as they could now build something which would be able to work and where it was clear what people could expect.
```
Comment (bart-blommaerts) [inline _posts/2016-12-12-Reactive-Programming-Spring-Reactor.md]: something, which

File: _posts/2016-12-12-Reactive-Programming-Spring-Reactor.md [hunk 4]
```diff
@@ -0,0 +1,525 @@
...
+This makes them easier to develop and to allow changes.
+They are significantly more tolerant of failure and when failure does occur they meet it with elegance rather than disaster.
+
+
+<a name="latency" />
+
+# Latency
+
+Latency is a real issue, the real physical distance of various components and services becomes more important with cloud based systems.
+This is also a very random number which is difficult to predict because it can depend on network congestion.
```
Comment (bart-blommaerts) [inline _posts/2016-12-12-Reactive-Programming-Spring-Reactor.md]: Latency is also a ...

File: _posts/2016-12-12-Reactive-Programming-Spring-Reactor.md [hunk 5]
```diff
@@ -0,0 +1,525 @@
...
+<a name ="contract" />
+
+# Contract
+
+Reactive is non-blocking and messages will never overflow the queue, see for the standard definition [http://www.reactive-streams.org/](http://www.reactive-streams.org/).
+
+>Created by Pivotal, Typesafe, Netflix, Oracle, Red Hat and others.
+
+The scope of Reactive Streams is to find a minimal set of interfaces, methods and protocols that will describe the necessary operations and entities to achieve the goal—asynchronous streams of data with non-blocking back pressure.
+Has been adopted for java 9.
```
Comment (bart-blommaerts) [inline _posts/2016-12-12-Reactive-Programming-Spring-Reactor.md]: Very short sentence ... :)

File: _posts/2016-12-12-Reactive-Programming-Spring-Reactor.md [hunk 6]
```diff
@@ -0,0 +1,525 @@
...
+This contract defines to send data `0 .. N`.
+[Publisher](http://www.reactive-streams.org/reactive-streams-1.0.0-javadoc/org/reactivestreams/Publisher.html) is an interface with a `subscribe()` method.
+[Subscriber](http://www.reactive-streams.org/reactive-streams-1.0.0-javadoc/org/reactivestreams/Subscriber.html) has 4 callback methods:
+`onSubscribe()`, `onNext()` (which can be called 0 to N times), `onComplete()` and `onError()`.
+The last two signals (complete and error) are terminal states, no further signals may occur and the subscriber's subscription is considered cancelled.
+
+What is important is the reverse flow and the back pressure.
+After subscribing, the subscriber gets a subscription which is a kind of 1 on 1 relationship between the subscriber and the publisher with 2 methods: `request` and `cancel`.
+
+* Request: this is the more important one, with this method the subscriber will ask the publisher to send me x messages (and not more), a so called 'pull'.
```
Comment (bart-blommaerts) [inline _posts/2016-12-12-Reactive-Programming-Spring-Reactor.md]: to send x messages (no me)

File: _posts/2016-12-12-Reactive-Programming-Spring-Reactor.md [hunk 7]
```diff
@@ -0,0 +1,525 @@
...
+
+<a name ="contract" />
+
+# Contract
+
+Reactive is non-blocking and messages will never overflow the queue, see for the standard definition [http://www.reactive-streams.org/](http://www.reactive-streams.org/).
+
+>Created by Pivotal, Typesafe, Netflix, Oracle, Red Hat and others.
+
+The scope of Reactive Streams is to find a minimal set of interfaces, methods and protocols that will describe the necessary operations and entities to achieve the goal—asynchronous streams of data with non-blocking back pressure.
```
Comment (bart-blommaerts) [inline _posts/2016-12-12-Reactive-Programming-Spring-Reactor.md]: Maybe also explain back pressure here (or link)? It is referenced later, but never explained.

File: _posts/2016-12-12-Reactive-Programming-Spring-Reactor.md [hunk 8]
```diff
@@ -0,0 +1,525 @@
...
+* Cancel: the subscription is being cancelled.
+
+Spring Reactor focuses on the publisher side of the reactive streaming, as this is the hardest to implement and to get right.
+It provides you with the tools to implement publishers in a back pressure way.
+
+The publisher is a provider of a potentially unbounded number of sequenced elements, publishing them according to the demand received from its Subscriber(s).
+
+# DIY Reactive Streams
+
+Implementing the Reactive Streams yourself is very hard to do, for Stephane Maldini this is the 4th or 5th attempt.
```
Comment (bart-blommaerts) [inline _posts/2016-12-12-Reactive-Programming-Spring-Reactor.md]: Implementing a Reactive Stream framework maybe?

File: _posts/2016-12-12-Reactive-Programming-Spring-Reactor.md [hunk 9]
```diff
@@ -0,0 +1,525 @@
...
+This is not really practical to work with, as most of the time we are only interested in a single user and not a stream of multiple results.
+When you would be building the method `findOneUser()` you also would not want to return an `Iterable` but just a single `User`.
+
+Also you will have to implement a subscriber to define the action to perform when the result is available.
+
+```
+rick.subscribe(new Subscriber<User>(){...});
+```
+
+Implementing this subscriber would not be that hard, because the specification has been made so that all complexity lies within the publishers side.
```
Comment (bart-blommaerts) [inline _posts/2016-12-12-Reactive-Programming-Spring-Reactor.md]: at the publishers side?

File: _posts/2016-12-12-Reactive-Programming-Spring-Reactor.md [hunk 10]
```diff
@@ -0,0 +1,525 @@
...
+Actually Reactor 3 was started as 2.5, but so many new features were added that the version had to be changed as well in order to reflect this.
+
+Since 3.0 Spring Reactor has been made more modular and consists of several components:
+<p style="text-align: center;">
+  <img alt="Spring Reactor Components" src="/img/reactive/spring-reactor-components.png">
+</p>
+
+* [Core](https://github.com/reactor/reactor-core) is the main library.
+Providing a non-blocking Reactive Streams foundation for the JVM both implementing a [Reactive Extensions](https://github.com/Reactive-Extensions) inspired API and efficient message-passing support.
+* [IPC](https://github.com/reactor/reactor-ipc): backpressure-ready components to encode, decode, send (unicast, multicast or request/response) and serve connections.
```
Comment (bart-blommaerts) [inline _posts/2016-12-12-Reactive-Programming-Spring-Reactor.md]: backpressure or back pressure?
Both are used in the article.

File: _posts/2016-12-12-Reactive-Programming-Spring-Reactor.md [hunk 11]
```diff
@@ -0,0 +1,525 @@
...
+
+# Reactive Types
+
+## Flux vs Observable
+<p style="text-align: center;">
+  <img alt="a Flux" src="/img/reactive/flux.png">
+</p>
+
+
+Also Observable is not implementing Reactive Streams Publisher which means that if you would like to use the Spring 5 save(Publisher<T>) you first have to convert the Observable to a Flowable.
```
Comment (bart-blommaerts) [inline _posts/2016-12-12-Reactive-Programming-Spring-Reactor.md]: Without "Also"?

File: _posts/2016-12-12-Reactive-Programming-Spring-Reactor.md [hunk 12]
```diff
@@ -0,0 +1,525 @@
...
+See [Observable vs Flowable](https://github.com/ReactiveX/RxJava/wiki/What's-different-in-2.0)
+
+This was too much noise for the Spring team, they are less dependant on Android developers so they could go all in with java 8.
+
+Flux is a Reactive Streams Publisher with basic flow operations.
+Where you start from a static method which will describe how the data will be generated, [just()](http://next.projectreactor.io/docs/core/release/api/reactor/core/publisher/Flux.html#just-T...-) is the simplest way
```
Comment (bart-blommaerts) [inline _posts/2016-12-12-Reactive-Programming-Spring-Reactor.md]: operations, where

File: _posts/2016-12-12-Reactive-Programming-Spring-Reactor.md [hunk 13]
```diff
@@ -0,0 +1,525 @@
...
+
+
+Interface CarRepository {
+    Flux<Car> findByColor(String color);
+}
+
+```
+This Flux will retrieve all cars which match the color "red" then those with the color "white" and finally "blue".
+So instead of just 3 elements, after this Flatmap we are going to have a lot more elements.
+This is all handled with backpressure in mind, fore example when the flatmap is busy merging data we will not ask for extra records
```
Comment (bart-blommaerts) [inline _posts/2016-12-12-Reactive-Programming-Spring-Reactor.md]: for example

File: _posts/2016-12-12-Reactive-Programming-Spring-Reactor.md [hunk 14]
```diff
@@ -0,0 +1,525 @@
...
+You can also make use of [Stepverifier](ttp://next.projectreactor.io/ext/docs/api/reactor/test/StepVerifier.html) to test Flux, Mono and any other kind of Reactive Streams Publisher.
+
+```
+
+@Test
+public void expectElementsWithThenComplete() {
+    expectSkylerJesseComplete(Flux.just(new User("swhite", null, null), new User("jpinkman", null, null)));
+}
+
+// Use StepVerifier to check that the flux parameter emits a User with "swhite" username and another one with "jpinkman" then completes successfully.
```
Comment (bart-blommaerts) [inline _posts/2016-12-12-Reactive-Programming-Spring-Reactor.md]: Maybe 2 code blocks with text in between and no comment?
Pure preference probably.

**Comment by bart-blommaerts** [comment]
Well done!

File: _posts/2016-12-12-Reactive-Programming-Spring-Reactor.md [hunk 15]
```diff
@@ -0,0 +1,527 @@
...
+* [Other Changes](#changes)
+* [RxJava](#rxjava)
+* [Spring Framework 5](#spring-5)
+* [Conclusion & Do It Yourself](#conclusion)
+
+<a name="join" />
+
+# Stephane Maldini @ JOIN 2016
+
+On 5 October 2016, we had the pleasure to welcome Stephane Maldini at our [JOIN event](https://ordina-jworks.github.io/conferences/2016/09/27/JOIN-2016.html)
```
Comment (yannickdeturck) [inline _posts/2016-12-12-Reactive-Programming-Spring-Reactor.md]: I would add a period at the end of the sentence.

File: _posts/2016-12-12-Reactive-Programming-Spring-Reactor.md [hunk 16]
```diff
@@ -0,0 +1,527 @@
...
+> You can rewatch his [talk](https://www.youtube.com/watch?v=RU0yQhfybDg) on on our [Channel](https://www.youtube.com/channel/UCsebfWdqV7LqNNDMDvCESIA) on Youtube.
+
+
+<a name="the-new-normal" />
+
+# The new normal that is not new
+
+It has been around for 30-40 years and boils down to [Event-Driven Programming](https://en.wikipedia.org/wiki/Event-driven_programming)
+What is new is "reactive motion bound to specification", this means that reactive programming is based on something solid, a specification and no longer some functional concepts.
+Namely the [Reactive Manifesto](http://www.reactivemanifesto.org/)
```
Comment (yannickdeturck) [inline _posts/2016-12-12-Reactive-Programming-Spring-Reactor.md]: Period at the end of the sentence.

File: _posts/2016-12-12-Reactive-Programming-Spring-Reactor.md [hunk 17]
```diff
@@ -0,0 +1,527 @@
...
+# Latency
+
+Latency is also a real issue, the real physical distance of various components and services becomes more important with cloud based systems.
+This is also a very random number which is difficult to predict because it can depend on network congestion.
+With [Zipkin](https://github.com/openzipkin/zipkin), you can measure this latency.
+
+The same latency can also exist within an application - between the different threads - although the impact will be less severe than between various components.
+
+Something needs to be done when latency becomes too big of an issue, especially if the receiver can not process enough.
+Too much data will fill up the buffer and can result, with an unbounded queue, to the infamous [OutOfMemoryException()](https://docs.oracle.com/javase/8/docs/api/java/lang/OutOfMemoryError.html)
```
Comment (yannickdeturck) [inline _posts/2016-12-12-Reactive-Programming-Spring-Reactor.md]: Period at the end of the sentence.

File: _posts/2016-12-12-Reactive-Programming-Spring-Reactor.md [hunk 18]
```diff
@@ -0,0 +1,527 @@
...
+While you won't run out of memory with a circular buffer, you risk losing messages as the oldest ones get overwritten.
+
+
+# Blocking
+
+One way to prevent out of memory exceptions is to use blocking.
+
+But this can be a real `poison pill`: when a queue is full, it will block a thread and as more and more queues get blocked your server will die a slow death.
```
Comment (yannickdeturck) [inline _posts/2016-12-12-Reactive-Programming-Spring-Reactor.md]: Maybe chance the code styling of poison pill to italic styling instead?

File: _posts/2016-12-12-Reactive-Programming-Spring-Reactor.md [hunk 19]
```diff
@@ -0,0 +1,527 @@
...
+<a name ="contract" />
+
+# Contract
+
+Reactive is non-blocking and messages will never overflow the queue, see for the standard definition [http://www.reactive-streams.org/](http://www.reactive-streams.org/).
+
+>Created by Pivotal, Typesafe, Netflix, Oracle, Red Hat and others.
+
+The scope of Reactive Streams is to find a minimal set of interfaces, methods and protocols that will describe the necessary operations and entities to achieve the goal—asynchronous streams of data with non-blocking [back-pressure](http://www.reactivemanifesto.org/glossary#Back-Pressure).
+With back-pressure, a consumer which can not handle the load of events sends towards it, can communicate this towards the upstream components so these can reduce the load.
```
Comment (yannickdeturck) [inline _posts/2016-12-12-Reactive-Programming-Spring-Reactor.md]: Think this needs to be changed to:
... a consumer **who** cannot...

File: _posts/2016-12-12-Reactive-Programming-Spring-Reactor.md [hunk 20]
```diff
@@ -0,0 +1,527 @@
...
+This contract defines to send data `0 .. N`.
+[Publisher](http://www.reactive-streams.org/reactive-streams-1.0.0-javadoc/org/reactivestreams/Publisher.html) is an interface with a `subscribe()` method.
+[Subscriber](http://www.reactive-streams.org/reactive-streams-1.0.0-javadoc/org/reactivestreams/Subscriber.html) has 4 callback methods:
+`onSubscribe()`, `onNext()` (which can be called 0 to N times), `onComplete()` and `onError()`.
+The last two signals (complete and error) are terminal states, no further signals may occur and the subscriber's subscription is considered cancelled.
+
+What is important is the reverse flow and the back-pressure.
+After subscribing, the subscriber gets a subscription which is a kind of 1 on 1 relationship between the subscriber and the publisher with 2 methods: `request` and `cancel`.
+
+* Request: this is the more important one, with this method the subscriber will ask the publisher to send x messages (and not more), a so called 'pull'.
```
Comment (yannickdeturck) [inline _posts/2016-12-12-Reactive-Programming-Spring-Reactor.md]: Maybe add some bold styling to Request and Cancel to make them stand out more.

File: _posts/2016-12-12-Reactive-Programming-Spring-Reactor.md [hunk 21]
```diff
@@ -0,0 +1,527 @@
...
+For example:
+
+```
+Publisher<User> rick = userRepository.findUser(“rick”);
+```
+
+Note that a publisher is returned instead of directly returning the entity.
+By doing so it does not block the subscribers when querying for the user and the publisher will produce the user when ready.
+
+But by using the specification as is, your publisher might produce 0,1 or N users, returning an `Iterable` as result.
```
Comment (yannickdeturck) [inline _posts/2016-12-12-Reactive-Programming-Spring-Reactor.md]: Add a space after the '0'

File: _posts/2016-12-12-Reactive-Programming-Spring-Reactor.md [hunk 22]
```diff
@@ -0,0 +1,527 @@
...
+
+# DIY Reactive Streams
+
+Implementing a Reactive Stream framework yourself is very hard to do, for Stephane Maldini this is the 4th or 5th attempt. 
+For Davik Karnok, the tech lead of RxJava, it is attempt 7 or 8.
+The main difficulty is to make it side effect free.
+
+For example:
+
+```
```
Comment (yannickdeturck) [inline _posts/2016-12-12-Reactive-Programming-Spring-Reactor.md]: For code samples in Java I would use the highlight syntax (examples see the Lagom blogpost). Everything looks better with colours! :-)

For example:
{% highlight java %}
public interface HelloService extends Service {
  ServiceCall<String, NotUsed, String> hello();
  
  ServiceCall<String, GreetingMessage, String> useGreeting();

  @Override
  default Descriptor descriptor() {
    return named("helloservice").with(
        restCall(Method.GET,  "/api/hello/:id",       hello()),
        restCall(Method.POST, "/api/hello/:id",       useGreeting())
      ).withAutoAcl(true);
  }
}
{% endhighlight %}

File: _posts/2016-12-12-Reactive-Programming-Spring-Reactor.md [hunk 23]
```diff
@@ -0,0 +1,527 @@
...
+
+Implementing this subscriber would not be that hard, because the specification has been made so that all complexity lies at the publishers side.
+
+Another issue is that you can only subscribe on the publisher, there are no other methods available like [map](http://martinfowler.com/articles/collection-pipeline/map.html), [flatmap](http://martinfowler.com/articles/collection-pipeline/flat-map.html), ...
+
+
+The other point is that when designing your own API you will also have to deal with the following issues:
+
+* Should work with RS TCK (otherwise it might not work with other libraries as well)
```
Comment (yannickdeturck) [inline _posts/2016-12-12-Reactive-Programming-Spring-Reactor.md]: Maybe make RS TCK clickable and refer to more information about it.

File: _posts/2016-12-12-Reactive-Programming-Spring-Reactor.md [hunk 24]
```diff
@@ -0,0 +1,527 @@
...
+</p>
+
+* [Core](https://github.com/reactor/reactor-core) is the main library.
+Providing a non-blocking Reactive Streams foundation for the JVM both implementing a [Reactive Extensions](https://github.com/Reactive-Extensions) inspired API and efficient message-passing support.
+* [IPC](https://github.com/reactor/reactor-ipc): back-pressure-ready components to encode, decode, send (unicast, multicast or request/response) and serve connections.
+Here you will find support for [Kafka](https://kafka.apache.org) and [Netty](http://netty.io).
+* [Addons](https://github.com/reactor/reactor-addons): Bridge to RxJava 1 or 2 Observable, Completable, Flowable, Single, Maybe, Scheduler, and also Swing/SWT Scheduler, Akka Scheduler.
+* [Reactive Streams Commons](https://github.com/reactor/reactive-streams-commons ) is the research project between Spring Reactor and RxJava as both teams had a lot of ideas they wanted to implement.
+Lots of effort was put in order to create real working, side-effect free operations.
+Map and Filtering for example are easy, but mergings, like Flatmap are hard to implement side-effect free.
```
Comment (yannickdeturck) [inline _posts/2016-12-12-Reactive-Programming-Spring-Reactor.md]: Could make Map, Filtering and Flatmap clickable and refer to more info.

File: _posts/2016-12-12-Reactive-Programming-Spring-Reactor.md [hunk 25]
```diff
@@ -0,0 +1,527 @@
...
+Having a proper implementation in the research project for these operations allowed the team to experiment and make it quite robust.
+This project contains Reactive-Streams compliant operators, which in turn are implemented by Spring Reactor and RxJava
```
Comment (yannickdeturck) [inline _posts/2016-12-12-Reactive-Programming-Spring-Reactor.md]: Period at the end.

File: _posts/2016-12-12-Reactive-Programming-Spring-Reactor.md [hunk 26]
```diff
@@ -0,0 +1,527 @@
...
+# Reactive Types
+
+## Flux vs Observable
+<p style="text-align: center;">
+  <img alt="a Flux" src="/img/reactive/flux.png">
+</p>
+
+
+Observable is not implementing Reactive Streams Publisher which means that if you would like to use the Spring 5 save(Publisher<T>) you first have to convert the Observable to a Flowable.
+See [Observable vs Flowable](https://github.com/ReactiveX/RxJava/wiki/What's-different-in-2.0)
```
Comment (yannickdeturck) [inline _posts/2016-12-12-Reactive-Programming-Spring-Reactor.md]: Mmm, some issue here? It doesn't show up as an hyperlink.

File: _posts/2016-12-12-Reactive-Programming-Spring-Reactor.md [hunk 27]
```diff
@@ -0,0 +1,527 @@
...
+       .subscribe(doWithResult);
+
+
+Interface CarRepository {
+    Flux<Car> findByColor(String color);
+}
+
+```
+This Flux will retrieve all cars which match the color "red" then those with the color "white" and finally "blue".
+So instead of just 3 elements, after this Flatmap we are going to have a lot more elements.
```
Comment (yannickdeturck) [inline _posts/2016-12-12-Reactive-Programming-Spring-Reactor.md]: three

File: _posts/2016-12-12-Reactive-Programming-Spring-Reactor.md [hunk 28]
```diff
@@ -0,0 +1,527 @@
...
+    .elapsed()
+    .subscribe()
+
+```
+This Mono will wait for 3 seconds on the "call" to Spring 4 or 2 seconds on that of Spring 5.
+The fastest result will be the one which will be outputted.
+
+
+The Mono has as advantage over an Observable Future of Java 8 that a Mono will only be triggered if you subscribe to it.
+While with an Observable the call to send() will execute the operation.
```
Comment (yannickdeturck) [inline _posts/2016-12-12-Reactive-Programming-Spring-Reactor.md]: `send()`

File: _posts/2016-12-12-Reactive-Programming-Spring-Reactor.md [hunk 29]
```diff
@@ -0,0 +1,527 @@
...
+}
+
+```
+
+# Debug
+
+When you use reactive libraries you will quickly realize that step debugging is hard especially when you try to read your stacktraces, there are a lot of recursive calls taking place.
+
+Before you invoke your operations you can enable an, expensive, debug mode.
+```
```
Comment (yannickdeturck) [inline _posts/2016-12-12-Reactive-Programming-Spring-Reactor.md]: Prepend with an <ENTER>

File: _posts/2016-12-12-Reactive-Programming-Spring-Reactor.md [hunk 30]
```diff
@@ -0,0 +1,527 @@
...
+    client.addListener(listener);
+    
+    emitter.setCancellation(() -> client.removeListener(listener));
+
+});
+
+```
+This `create()` allows you to bridge 1 result, which will be returned somewhere in the future, to a Mono.
+
+If you add a Kafka call, for example, where they have this callback so one can return onSuccess and onError you can use `Mono.create()`:
```
Comment (yannickdeturck) [inline _posts/2016-12-12-Reactive-Programming-Spring-Reactor.md]: `onSuccess` and `onError`

File: _posts/2016-12-12-Reactive-Programming-Spring-Reactor.md [hunk 31]
```diff
@@ -0,0 +1,527 @@
...
+
+Reactor: a Simpler API, the entire framework just fits in 1 jar: reactor-core jar.
+Flux and Mono live in the `reactor.com.publisher` package, `reactor.core.scheduler` contains the FIFO task executor.
+
+By default the `Publisher` and `Subscriber` will use the same thread.
+With `publishOn()` the publisher can force the subscriber to use a different thread, while the subscriber can do the same with `subscribeOn()`.
+
+For Reactor 3.x there will be more focus on the [javadoc](http://projectreactor.io/core/docs/api), as this has been lagging behind the new developments.
+
+I understand, as we developers are so good with keeping our documentation up to date ;-)
```
Comment (yannickdeturck) [inline _posts/2016-12-12-Reactive-Programming-Spring-Reactor.md]: Would rephrase this because it sounds a bit odd.

File: _posts/2016-12-12-Reactive-Programming-Spring-Reactor.md [hunk 32]
```diff
@@ -0,0 +1,527 @@
...
+
+<a name="rxjava" />
+
+# RxJava
+
+Why Reactor when there’s already [RxJava2](https://github.com/ReactiveX/RxJava/wiki/What's-different-in-2.0)?
+
+RxJava2 is java 6 while for Reactor the Spring team decided to go all in and focus only on java 8.
```
Comment (yannickdeturck) [inline _posts/2016-12-12-Reactive-Programming-Spring-Reactor.md]: **J**ava

File: _posts/2016-12-12-Reactive-Programming-Spring-Reactor.md [hunk 33]
```diff
@@ -0,0 +1,527 @@
...
+This means that you can make use of all the new and fancy java 8 features.
```
Comment (yannickdeturck) [inline _posts/2016-12-12-Reactive-Programming-Spring-Reactor.md]: **J**ava
